### PR TITLE
feat(sdk): add missing utilities for `fs` module

### LINKED ISSUE #4838: Missing some essential `fs` utilities
I am missing:
* `fs.isdir`: returns `true` if a path is a directory and exists
* `fs.stat`, `fs.lstat`
* `fs.chmod`
* `fs.absolute(...dirs)` (resolves an absolute path from a sequence of dir specs)
* `fs.relative(from, to)`
* `fs.extname`

### PR DESCRIPTION
**Pull request by garysassano**
Closes #4838

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [ ] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: libs/wingsdk/src/fs/fs.ts [hunk 1]
```diff
@@ -127,13 +153,13 @@ export class Util {
...
+    return normalPath(nodePath.resolve(...paths));
   }
 
   /**
    * Check if the path exists.
-   * @param p The path to evaluate.
+   * @param path The path to evaluate.
    * @returns `true` if the path exists, `false` otherwise.
    */
   public static exists(p: string): boolean {
```
Comment (eladb) [inline libs/wingsdk/src/fs/fs.ts:165]:
```suggestion
  public static exists(path: string): boolean {
```

File: libs/wingsdk/src/fs/fs.ts [hunk 2]
```diff
@@ -314,7 +340,7 @@ export class Util {
 
   /**
    * Remove files and directories (modeled on the standard POSIX `rm`utility). Returns `undefined`.
-   * @param p The path to the file or directory you want to remove.
+   * @param path The path to the file or directory you want to remove.
    */
   public static remove(p: string, opts?: RemoveOptions): void {
```
Comment (eladb) [inline libs/wingsdk/src/fs/fs.ts:345]:
```suggestion
  public static remove(path: string, opts?: RemoveOptions): void {
```

File: libs/wingsdk/src/fs/fs.ts [hunk 3]
```diff
@@ -323,10 +349,124 @@ export class Util {
     });
   }
 
+  /**
+   * Checks if the given path is a directory and exists.
+   * @param path The path to check.
+   * @returns `true` if the path is an existing directory, `false` otherwise.
+   */
+  public static isDir(p: string): boolean {
```
Comment (eladb) [inline libs/wingsdk/src/fs/fs.ts:357]: Please check everywhere that the argument matches the docstring

File: libs/wingsdk/src/fs/fs.ts [hunk 4]
```diff
@@ -323,10 +349,124 @@ export class Util {
...
+      return false;
+    }
+  }
+
+  /**
+   * Gets the stats of the given path.
+   * @param path The path to get stats for.
+   * @returns The stats of the path, formatted as a `Metadata` object.
+   */
+  public static metadata(p: string): Metadata {
```
Comment (eladb) [inline libs/wingsdk/src/fs/fs.ts:370]: please rename to `stat`

File: libs/wingsdk/src/fs/fs.ts [hunk 5]
```diff
@@ -323,10 +349,124 @@ export class Util {
...
+    return this._metadata(fs.statSync(p));
+  }
+
+  /**
+   * Gets the stats of the given path without following symbolic links.
+   * @param path The path to get stats for.
+   * @returns The stats of the path, formatted as a `Metadata` object.
+   */
+  public static symlinkMetadata(p: string): Metadata {
```
Comment (eladb) [inline libs/wingsdk/src/fs/fs.ts:379]: please rename to `lstat`

File: libs/wingsdk/src/fs/fs.ts [hunk 6]
```diff
@@ -323,10 +349,124 @@ export class Util {
...
+    return this._metadata(fs.lstatSync(p));
+  }
+
+  /**
+   * Set the permissions of the file, directory, etc.
+   * Expects a permission string like `"755"` or `"644"`.
+   * @param path The path of the file or directory.
+   * @param permissions The mode to set as a string.
+   */
+  public static setPermissions(p: string, permissions: string): void {
```
Comment (eladb) [inline libs/wingsdk/src/fs/fs.ts:389]: please rename to `chmod`

File: libs/wingsdk/src/fs/fs.ts [hunk 7]
```diff
@@ -323,10 +349,124 @@ export class Util {
...
+
+  /**
+   * Extracts the extension (without the leading dot) from the path, if possible.
+   * @param path The path to get extension for.
+   * @returns The file extension without the leading dot, or `nil` if:
+   *          - The file name starts with a dot (hidden files).
+   *          - There is no dot in the file name.
+   *          - The dot is the last character in the file name.
+   */
+  public static extension(p: string): string | undefined {
```
Comment (eladb) [inline libs/wingsdk/src/fs/fs.ts:401]: please rename to `extname`

File: libs/wingsdk/src/fs/fs.ts [hunk 8]
```diff
@@ -323,10 +349,124 @@ export class Util {
...
+   * Creates a symbolic link.
+   * @param target The path to the target file or directory.
+   * @param path The path to the symbolic link to be created.
+   * @param type The type of the target. It can be `"file"`, `"dir"`, or `"junction"` (Windows only).
+   *             Defaults to `"file"` if not specified.
+   */
+  public static symlink(
+    target: string,
+    p: string,
+    type: "file" | "dir" | "junction" = "file"
```
Comment (eladb) [inline libs/wingsdk/src/fs/fs.ts:416]: please use an enum

File: libs/wingsdk/src/fs/fs.ts [hunk 9]
```diff
@@ -323,10 +349,124 @@ export class Util {
...
+
+  /**
+   * Returns the type of the file based on the given `fs.Stats` object.
+   * @param stats The `fs.Stats` object.
+   * @returns The type of the file.
+   */
+  private static _fileType(stats: fs.Stats): FileType {
+    switch (true) {
+      case stats.isFile():
+        return "File";
```
Comment (eladb) [inline libs/wingsdk/src/fs/fs.ts:452]:
```suggestion
        return FileType.File;
```

File: libs/wingsdk/src/fs/fs.ts [hunk 10]
```diff
@@ -323,10 +349,124 @@ export class Util {
...
+  /**
+   * Returns the `Metadata` object based on the given `fs.Stats` object.
+   * @param stats The `fs.Stats` object.
+   * @returns The `Metadata` object.
+   */
+  private static _metadata(stats: fs.Stats): Metadata {
+    return {
+      fileType: this._fileType(stats),
+      size: stats.size,
+      permissions: this._formatPermissions(stats.mode),
```
Comment (eladb) [inline libs/wingsdk/src/fs/fs.ts:437]:
```suggestion
      mode: this._formatPermissions(stats.mode),
```

**Review by eladb — CHANGES_REQUESTED** [review]
i prefer that we retain POSIX terminology here

File: libs/wingsdk/src/fs/fs.ts (hunk 4 above)
Comment (eladb) [reply libs/wingsdk/src/fs/fs.ts:370]: ok, I am okay with the rust terminology, so ignore these comments...

File: libs/wingsdk/src/fs/fs.ts (hunk 5 above)
Comment (eladb) [reply libs/wingsdk/src/fs/fs.ts:379]: ignore

File: libs/wingsdk/src/fs/fs.ts (hunk 6 above)
Comment (eladb) [reply libs/wingsdk/src/fs/fs.ts:389]: ignore

**Review by eladb — APPROVED** [review]
ok, I am okay with the rust terminology, so ignore these comments...
i saw @Chriscbr's comment on the issue.

ship it (with a few tweaks)

**Comment by Chriscbr** [comment]
Thanks @garysassano !

**Pull request merged**